File: bend2/effs/tcp_connect_poll.js
// TCP
// ===

// TCP.connect with a deadline, as tcp_connect.js: EINPROGRESS parks the
// computation on the socket becoming writable and on the clock. Woken,
// SO_ERROR says whether the connect failed, and a second connect whether
// it is done (EISCONN) or still going (EALREADY); the clock past, it is
// ETIMEDOUT.
function tcp_connect_poll(host, port, ms, k) {
  const sys = io_sys();
  const at = io_addr(host, Number(port));
  if (at === null) {
    return io_fail(22);
  }
  const fd = sys.socket(2, 1, 0);
  if (fd < 0) {
    return io_fail(sys.errno());
  }
  const end = (code) => {
    if (code !== 0) {
      sys.close(fd);
      return io_fail(code);
    }
    return io_done(fd);
  };
  const error = () => {
    const v = new Int32Array([0]);
    const l = new Uint32Array([4]);
    return sys.getsockopt(fd, sys.mac ? 0xffff : 1, sys.mac ? 0x1007 : 4,
      sys.ptr(v), sys.ptr(l)) < 0 ? sys.errno() : v[0];
  };
  const set = sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800));
  const ok = set >= 0 && sys.connect(fd, sys.ptr(at), 16) >= 0;
  const code = ok ? 0 : sys.errno();
  const going = sys.mac ? [36, 37] : [115, 114];
  if (code !== going[0]) {
    return end(code);
  }
  const due = performance.now() + Math.max(Number(ms), 1);
  const more = () => {
    const err = error();
    if (err !== 0) {
      return end(err);
    }
    const again = sys.connect(fd, sys.ptr(at), 16) >= 0 ? 0 : sys.errno();
    if (!going.includes(again)) {
      return end(again === (sys.mac ? 56 : 106) ? 0 : again);
    }
    if (performance.now() >= due) {
      return end(sys.mac ? 60 : 110);
    }
    io_park_on(fd, true, k, more, due);
    return undefined;
  };
  io_park_on(fd, true, k, more, due);
  return undefined;
}

File: bend2/effs/tcp_connect_poll.c
// TCP
// ===

// TCP.connect with a deadline, to a dotted IPv4 address or an IPv6 one:
// the socket is non-blocking for life, a
// connect that is not done at once parks on the socket becoming writable
// and on the clock, and whichever comes first ends it. Writable, SO_ERROR
// says how the connect went; the clock first, it is ETIMEDOUT and the
// socket is closed. w->made holds the descriptor.
#include <netinet/tcp.h>
#include <poll.h>

static Term tcp_connect_poll_end(Env e, IoWork* w, int err) {
  int fd = (int)w->made;
  if (err != 0 && fd >= 0) {
    close(fd);
  }
  free(w->data);
  return err != 0 ? io_fail(e, (u32)err, NULL) : io_done(e, io_hand(fd));
}

static Term tcp_connect_poll_more(Env e, IoWork* w) {
  int           fd = (int)w->made;
  struct pollfd p  = { fd, POLLOUT, 0 };
  if (poll(&p, 1, 0) > 0) {
    int       err = 0;
    socklen_t len = sizeof(err);
    if (getsockopt(fd, SOL_SOCKET, SO_ERROR, &err, &len) != 0) {
      err = errno;
    }
    return tcp_connect_poll_end(e, w, err);
  }
  u64 at = io_wait_time(w);
  return io_tick() < at ? io_wait_on(w, fd, POLLOUT, at, tcp_connect_poll_more)
    : tcp_connect_poll_end(e, w, ETIMEDOUT);
}

Term tcp_connect_poll_run(Env e, Term* f, IoWork* w) {
  struct sockaddr_storage at;
  socklen_t               len = 0;
  w->data = io_cstr(e, f[0], &w->size);
  w->made = -1;
  if (io_nul(w->data, w->size) || io_sys_sa(w->data, (u32)f[1], &at, &len) < 0) {
    return tcp_connect_poll_end(e, w, EINVAL);
  }
  int fd = socket(at.ss_family, SOCK_STREAM, 0);
  if (fd < 0) {
    return tcp_connect_poll_end(e, w, errno);
  }
  w->made = fd;
  int one = 1;
  setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
  if (fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    return tcp_connect_poll_end(e, w, errno);
  }
  if (connect(fd, (struct sockaddr*)&at, len) == 0) {
    return tcp_connect_poll_end(e, w, 0);
  }
  if (errno != EINPROGRESS) {
    return tcp_connect_poll_end(e, w, errno);
  }
  u64 ms = (u64)f[2];
  return io_wait_on(w, fd, POLLOUT, io_tick() + (ms == 0 ? 1 : ms) * 1000000ull,
    tcp_connect_poll_more);
}

static void __attribute__((constructor)) tcp_connect_poll_use(void) {
  io_eff(CID_TCP_CONNECT_POLL, tcp_connect_poll_run, 0);
}
